File: src/data_algorithm_mst/Prim.java
package data_algorithm_mst;


import data_algorithm_dijkstra.Edge;

import java.util.PriorityQueue;

/**
 * Created by parkdongjo on 2017. 9. 19..
 */
public class Prim {
    public static void main(String[] args) {

        Integer[][] graph = { { null, 1, 6, 7, 5 },
                            { 1, null, 2, 9, 8 },
                            { 6, 2, null, 3, 10 },
                            { 7, 9, 3, null, 4 },
                            { 5, 8, 10, 4, null } };

        printMSTByPrim(graph);
    }

    public static void printMSTByPrim(Integer[][] graph) {
        int start = 0;
        int count = 0;
        PriorityQueue<Edge> queue = new PriorityQueue<Edge>();
        Edge[] MST = new Edge[graph[start].length -1];


        MST = prim(graph,start,queue,MST,count);


        for(int i = 0; i <MST.length ; i++) {
            System.out.println(MST[i].getWeight());
        }
    }

    public static Edge[] prim(Integer[][] graph, int start, PriorityQueue<Edge> queue, Edge[] MST, int count) {

        if(count == graph[0].length -1) {
            return MST;
        }

        for(int i=0; i<graph[start].length; i++){
            if(graph[start][i] != null) {
                queue.add(new Edge(start,i,graph[start][i]));
            }
        }
        Edge edge = queue.poll();
        int next = edge.getEnd();
        graph[start][next] = null;
        graph[next][start] = null;
        MST[count++] = edge;


        return prim(graph,next,queue,MST,count);
    }

}

File: src/data_algorithm_mst/Prim.kt
package data_algorithm_mst

import java.util.*
import kotlin.Comparator

val NODE_MAX = 1001

fun main(args: Array<String>) {
    var graph = arrayOf(
            intArrayOf(NODE_MAX, 1, 6, 7, 5),
            intArrayOf(1, NODE_MAX, 2, 9, 8),
            intArrayOf(6, 2, NODE_MAX, 3, 10),
            intArrayOf(7, 9, 3, NODE_MAX, 4),
            intArrayOf(5, 8, 10, 4, NODE_MAX)
            )
    var start = 0
    var cp = Comp()
    var pQueue = PriorityQueue<Edge>(cp)
    var MST = mutableListOf<Edge>()

    prim(graph, start, pQueue, MST)

    for (e in MST) {
        println(e.weight)
    }
}

fun prim(graph: Array<IntArray>,
         start: Int,
         pQueue: PriorityQueue<Edge>,
         MST: MutableList<Edge>): MutableList<Edge> {

    if(MST.size == graph[0].size) return MST

    for ((i, w) in graph[start].withIndex()) {
        if (w != NODE_MAX) {
            pQueue.add(Edge(start, i, w))
        }
    }
    var edge = pQueue.poll()
    var next = edge.end
    graph[start][next] = NODE_MAX
    graph[next][start] = NODE_MAX
    MST.add(edge)

    return prim(graph, next, pQueue, MST)
}

class Comp: Comparator<Edge> {
    override fun compare(o1: Edge?, o2: Edge?): Int {
        return if (o1!!.weight > o2!!.weight) 1 else -1
    }
}